plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -457,20 +460,45 @@ throwingDischarged
     -> CekM uni fun s x
 throwingDischarged l t = throwingWithCause l t . Just . dischargeCekValue
 
-instance ThrowableBuiltins uni fun => MonadError (CekEvaluationException NamedDeBruijn uni fun) (CekM uni fun s) where
+instance ThrowableBuiltins uni fun =>
+        MonadError (CekEvaluationException NamedDeBruijn uni fun) (CekM uni fun s) where
     -- See Note [Throwing exceptions in ST].
     throwError = CekM . throwM
 
     -- See Note [Catching exceptions in ST].
-    a `catchError` h = CekM . unsafeIOToST $ aIO `catch` hIO where
+    catchError
+        :: forall a.
+           CekM uni fun s a
+        -> (CekEvaluationException NamedDeBruijn uni fun -> CekM uni fun s a)
+        -> CekM uni fun s a
+    a `catchError` h =
+        -- Here in addition to catching 'CekEvaluationException' we also catch common GHC exceptions
+        -- in case one of them somehow gets triggered during script execution (which would be a bug
+        -- on our side). We could probably use @enclosed-exceptions@, but spawning a thread per
+        -- script may be considered expensive, so we do the simplest thing possible.
+        CekM . unsafeIOToST $ aIO `catches`
+            [ Handler hIO
+            , panicHandler @IOError
```

Comment:
Why these particular exceptions? Where do you get the list from?

Reply:
I just looked at all the instances of the `Exception` class and picked those that seemed like they would definitely be useful to catch. We could use [this](https://next.fpcomplete.com/blog/async-exception-handling-haskell/) instead to catch all exceptions whose type indicated that they're sync ones:

![image](https://github.com/user-attachments/assets/748d5ba1-1eaa-47b9-8f4e-2184f5d145b1)

But the type may lie unfortunately.

I'll ask about `enclosed-exceptions` though, it's the right approach, it's just costly.